GiniVision/Classes/GiniQRCodeDocument.swift

```diff
@@ -0,0 +1,162 @@
+        }
+    }
+    
+    fileprivate func generateQRCodeImage(from data: Data) -> UIImage? {
+        let filter = CIFilter(name: "CIQRCodeGenerator")
+        
+        filter?.setValue(data, forKey: "inputMessage")
+        filter?.setValue("Q", forKey: "inputCorrectionLevel")
+        
+        if let outputImage = filter?.outputImage {
+            return UIImage(ciImage: outputImage.applying(CGAffineTransform(scaleX: 2, y: 2)))
+        }
+        
+        return nil
+    }
+}
+
+// MARK: Equatable
+
+extension GiniQRCodeDocument {
+    public override func isEqual(_ object: Any?) -> Bool {
+        if let object = object as? GiniQRCodeDocument {
+            return self.scannedString == object.scannedString
+        }
+        return false
+    }
+}
+
+// MARK: Extractions
+
+extension GiniQRCodeDocument {
+    fileprivate func extractParameters(from string: String) -> [String: String] {
+        switch qrCodeFormat {
+        case .some(.bezahlcode):
+            return extractParameters(fromBezhalCodeString: string)
+        case .some(.epc06912):
+            return extractParameters(fromEPC06912CodeString: string)
+        case .none:
+            return [:]
+        }
+    }
+    
+    fileprivate func extractParameters(fromBezhalCodeString string: String) -> [String: String] {
+        var parameters: [String: String] = [:]
+        
+        if let queryParameters = URL(string: string)?.queryParameters {
+            
+            if let bic = queryParameters["bic"] as? String {
+                parameters["bic"] = bic
+            }
+            if let paymentRecipient = queryParameters["name"] as? String {
+                parameters["paymentRecipient"] = paymentRecipient
+            }
+            if let iban = queryParameters["iban"] as? String,
+                IBANValidator().isValid(iban: iban) {
+                parameters["iban"] = iban
+            }
+            if let paymentReference = queryParameters["reason"] as? String ??
+                queryParameters["reason1"] as? String {
+                parameters["paymentReference"] = paymentReference
+            }
+            if let amount = queryParameters["amount"] as? String,
+                let amountNormalized = normalize(amount: amount,
+                                                 currency: queryParameters["currency"] as? String ?? "EUR") {
+                parameters["amountToPay"] = amountNormalized
+            }
+        }
+        
+        return parameters
+    }
+    
+    fileprivate func extractParameters(fromEPC06912CodeString string: String) -> [String: String] {
+        let lines = string.splitlines
+        var parameters: [String: String] = [:]
+        
+        if !lines[4].isEmpty {
+            parameters["bic"] = lines[4]
+        }
+        
+        if !lines[5].isEmpty {
+            parameters["paymentRecipient"] = lines[5]
+        }
+        
+        if !lines[9].isEmpty {
```

Comment:
How can you be sure all those 9 lines exist?

Reply:
Because I checked it when the document is validated.